cpp/src/io/orc/stats_enc.cu

```diff
@@ -0,0 +1,380 @@
+        break;
+      default: break;
+      }
+    }
+    stats_pos = WarpReducePos32(stats_len, t);
+    if ((t & 0x1f) == 0x1f) {
+      scratch_red[t >> 5] = stats_pos;
+    }
+    __syncthreads();
+    if (t < 32) {
+      scratch_red[t] = WarpReducePos32(scratch_red[t], t);
+    }
+    __syncthreads();
+    if (t >= 32) {
+      stats_pos += scratch_red[(t >> 5) - 1];
+    }
+    stats_pos += stats_size;
+    if (idx < statistics_count) {
+      groups[idx].start_chunk = stats_pos - stats_len;
+      groups[idx].num_chunks = stats_len;
+    }
+    __syncthreads();
+    if (t == 1023) {
+      stats_size = stats_pos;
+    }
+  }
+}
+
+
+struct stats_state_s {
+  uint8_t *base;              ///< Output buffer start
+  uint8_t *end;               ///< Output buffer end
+  statistics_chunk ck;
+  statistics_merge_group grp;
+  stats_column_desc col;
+  // ORC stats
+  uint64_t numberOfValues;
+  uint8_t hasNull;
+};
+
+
+// Protobuf varint encoding for unsigned int
+__device__ inline uint8_t *pb_encode_uint(uint8_t *p, uint64_t v) {
+  while (v > 0x7f) {
+    *p++ = ((uint32_t)v | 0x80);
+    v >>= 7;
+  }
+  *p++ = v;
+  return p;
+}
+
+// Protobuf field encoding for unsigned int
+__device__ inline uint8_t *pb_put_uint(uint8_t *p, uint32_t id, uint64_t v) {
+  p[0] = id * 8 + PB_TYPE_VARINT; // NOTE: Assumes id < 16
+  return pb_encode_uint(p + 1, v);
+}
+
+// Protobuf field encoding for signed int
+__device__ inline uint8_t *pb_put_int(uint8_t *p, uint32_t id, int64_t v) {
+  int64_t s = (v < 0);
+  return pb_put_uint(p, id, (v ^ -s) * 2 + s);
+}
+
+// Protobuf field encoding for 'packed' unsigned int (single value)
+__device__ inline uint8_t *pb_put_packed_uint(uint8_t *p, uint32_t id, uint64_t v) {
+  uint8_t *p2 = pb_encode_uint(p + 2, v);
+  p[0] = id * 8 + PB_TYPE_FIXEDLEN;
+  p[1] = static_cast<uint8_t>(p2 - (p + 2));
+  return p2;
+}
+
+// Protobuf field encoding for binary/string
+__device__ inline uint8_t *pb_put_binary(uint8_t *p, uint32_t id, const void *bytes, uint32_t len) {
+  p[0] = id * 8 + PB_TYPE_FIXEDLEN;
+  p = pb_encode_uint(p + 1, len);
+  memcpy(p, bytes, len);
+  return p + len;
+}
+
+// Protobuf field encoding for 64-bit raw encoding (double)
+__device__ inline uint8_t *pb_put_fixed64(uint8_t *p, uint32_t id, const void *raw64) {
+  p[0] = id * 8 + PB_TYPE_FIXED64;
+  memcpy(p + 1, raw64, 8);
+  return p + 9;
+}
+
+
+
+/**
+ * @brief Encode statistics in ORC protobuf format
+ *
+ * @param[in,out] groups Statistics merge groups
+ * @param[in,out] chunks Statistics data
+ * @param[in] statistics_count Number of statistics buffers
+ *
+ **/
+// blockDim {128,1,1}
+__global__ void __launch_bounds__(128)
+gpuEncodeStatistics(uint8_t *blob_bfr, statistics_merge_group *groups, const statistics_chunk *chunks,
+                    uint32_t statistics_count) {
+  __shared__ __align__(8) stats_state_s state_g[4];
+  uint32_t t = threadIdx.x & 0x1f;
```

Comment:
As above (warp size dependencies)

Reply:
I'll try to make it look like the warp size is abstracted. I don't think it's entirely a coincidence that the warp size matches the register size (to me, 'abstracting' the warp size just makes it more difficult to understand what's really happening)

Reply:
FWIW, G80 had a mode where the warp size was only 16 threads (before it was called a warp). It was never exposed, but it is a precedent.